### PR TITLE
Eagerly change the listeners count

### PR DESCRIPTION
**Pull request by fourlastor**
While working with `RCTEventEmitter` I noticed that if an event is emitted before `_listenerCount` is updated, it will not go through because the listeners count hasn't been updated. Moving the count update before the invokation of `startObserving` and `stopObserving` fixes the issue. Same way if you remove the last listener and an event is fired before the count is updated (while it shouldn't be fired).

**Test plan (required)**

An easy test to demonstrate it is to implement `startObserving` to synchronously fire an event. Without the change, a warning is thrown, with the change, the event is fired. Not very strong on Obj-C here and I didn't know how to mock out the native stuff. Would be glad to write a failing unit test tho :)

### CODE REVIEW

File: React/Modules/RCTEventEmitter.m [hunk 1]
```diff
@@ -78,21 +78,22 @@ - (void)dealloc
     RCTLogError(@"`%@` is not a supported event type for %@. Supported events are: `%@`",
                 eventName, [self class], [[self supportedEvents] componentsJoinedByString:@"`, `"]);
   }
-  if (_listenerCount == 0) {
+  _listenerCount++;
+  if (_listenerCount == 1) {
```
Comment (javache) [inline React/Modules/RCTEventEmitter.m]: nit: `(_listenerCount > 0)`

File: React/Modules/RCTEventEmitter.m [hunk 2]
```diff
@@ -78,21 +78,22 @@ - (void)dealloc
...
 }
 
 RCT_EXPORT_METHOD(removeListeners:(NSInteger)count)
 {
   if (RCT_DEBUG && count > _listenerCount) {
     RCTLogError(@"Attempted to remove more %@ listeners than added", [self class]);
   }
-  if (count == _listenerCount) {
+  _listenerCount -= count;
+  _listenerCount = (_listenerCount <= 0) ? 0 : _listenerCount;
```
Comment (javache) [inline React/Modules/RCTEventEmitter.m]: nit: `_listenerCount = MAX(_listenerCount  - _count, 0)`

File: React/Modules/RCTEventEmitter.m (hunk 1 above)
Comment (fourlastor) [inline React/Modules/RCTEventEmitter.m]: that would trigger this every time you add a listener, while you want it to be triggered only the first time as far as I understood

File: React/Modules/RCTEventEmitter.m (hunk 2 above)
Comment (fourlastor) [inline React/Modules/RCTEventEmitter.m]: do I need to import anything? fairly new to objective-c :smile: I usually work on the Android side of things

Comment (fourlastor) [inline React/Modules/RCTEventEmitter.m]: aa44160 seems like not from what I saw :)

**Pull request closed without merging**